modules/docereeBidAdapter.js

```diff
@@ -0,0 +1,105 @@
+    const { placementId } = bid.params;
+    return !!placementId
+  },
+  buildRequests: (validBidRequests) => {
+    const serverRequests = [];
+    const { data } = config.getConfig('doceree.user')
+    const { page, domain, token } = config.getConfig('doceree.context')
+    const encodedUserInfo = window.btoa(encodeURIComponent(JSON.stringify(data)))
+
+    validBidRequests.forEach(function(validBidRequest) {
+      const { publisherUrl, placementId } = validBidRequest.params;
+      const url = publisherUrl || page
+      let queryString = '';
+      queryString = utils.tryAppendQueryString(queryString, 'id', placementId);
+      queryString = utils.tryAppendQueryString(queryString, 'publisherDomain', domain);
+      queryString = utils.tryAppendQueryString(queryString, 'pubRequestedURL', encodeURIComponent(url));
+      queryString = utils.tryAppendQueryString(queryString, 'loggedInUser', encodedUserInfo);
+      queryString = utils.tryAppendQueryString(queryString, 'currentUrl', url);
+      queryString = utils.tryAppendQueryString(queryString, 'prebidjs', true);
+      queryString = utils.tryAppendQueryString(queryString, 'token', token);
+      queryString = utils.tryAppendQueryString(queryString, 'requestId', validBidRequest.bidId);
+
+      serverRequests.push({
+        method: 'GET',
+        url: END_POINT + '/v1/adrequest?' + queryString
+      })
+    })
+    return serverRequests;
+  },
+  interpretResponse: (serverResponse, request) => {
+    const responseJson = serverResponse ? serverResponse.body : {};
+    const placementId = responseJson.DIVID;
+    const bidResponse = {
+      width: Number(responseJson.width),
+      height: Number(responseJson.height),
+      requestId: responseJson.guid,
+      netRevenue: true,
+      ttl: 30,
+      cpm: responseJson.cpmBid,
+      currency: responseJson.currency,
+      mediaType: 'banner',
+      creativeId: placementId,
+      meta: {
+        advertiserDomains: [responseJson.advertiserDomain]
+      }
+    };
+
+    if (responseJson.creativeType && responseJson.creativeType !== 'banner') {
+      if (responseJson.sourceHTML) {
+        responseJson.sourceHTML = responseJson.sourceHTML.replace(/[\uE000-\uF8FF]/g, '');
+        let creative = processHTMLAndRender(responseJson, responseJson.sourceHTML)
+        bidResponse.ad = `<div id="${placementId}">${creative}</div>`
+      } else {
+        bidResponse.ad = `<iframe id="ifrm" src="${responseJson.sourceURL}" onload="${registerImpression(responseJson)}" style="border:none;margin:0"></iframe>`
+      }
+    } else {
+      if (responseJson.ctaLink) {
+        bidResponse.ad = `<a href="${responseJson.ctaLink}" target="_blank"><img onload="${registerImpression(responseJson)}" src="${responseJson.sourceURL}" /></a>`;
+      } else {
+        bidResponse.ad = `<div id="${placementId}"><img onload="${registerImpression(responseJson)}" src="${responseJson.sourceURL}" /></div>`;
+      }
+    }
+    return [bidResponse];
+  }
+};
+
```

Comment:
There seems to be a few operations on the creative on the client-side? I also don't really like the idea of re-constructing DOM elements on the client-side this way and then rendering. Is the purpose of all this for analytics or future communication with another adapter?

Reply:
did the changes